### ISSUE TITLE
Change the recommended plugin for TypeScript

### ISSUE THREAD

**Issue by mnapoli** [serverless/serverless#10700]
@arielweinberger mentioned in #10691 that the [serverless-plugin-typescript](https://github.com/serverless/serverless-plugin-typescript) might have better alternatives, and that we could recommend another plugin by default in examples.

I would love to get feedback from TypeScript users on which plugins they would recommend using by default: is there a consensus or not?

Please post which TypeScript plugin you think is best and why, and let's see if there's a clear option.

cc @fredericbarthelet

**Comment by pl0xy** [comment]
I use Typescript, and have done exclusively, with lambdas (even contributed to [@definitylyTyped/aws-lambda](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/aws-lambda) for some time now. Always using [@ServerlessWebpack](https://github.com/serverless-heaven/serverless-webpack). The setup was initially really confusing but it only needed done once and has actually gotten simpler over time as webpack its self has improved.

I saw [serverless-plugin-typescript](https://github.com/serverless/serverless-plugin-typescript) just earlier today, and I looed at it and it looks really interesting. but right now, webpack performs more than jsut acting as a means of transpiling typescript and doing type checking. it minifies, tree shakes, strips comments etc. I think for those people who use Serverless-Offline i am sue it will be of great help. or people whose applications are okay without being as tiny as possible for example. But unless i am missing something. I'd still have to run webpack to get those features anyway.

For sure am liking the move towards native support of typescript though, Typescript Bestscript

**Comment by pl0xy** [comment]
> Hello ! I have used [serverless-esbuild](https://github.com/floydspace/serverless-esbuild) on many occasions and found it to be the best option.

**Comment by suresh430** [comment]
From three installation routes, typeScript can be installed on how we intend to use it: a NuGet package,npm module, Visual Studio Extension. If you are using Node.js, you want the npm version. If you are using MSBuild you want the NuGet package or Visual Studio extension.

**Comment by arielweinberger** [comment]
@pl0xy 
Indeed, I find webpack to be very powerful not just for transpilation but for reducing bundle size and tree shaking as well. I'll try out `serverless-esbuild` as well (so far been using `serverless-bundle`).

What I will say, is that `serverless-plugin-typescript` unfrotunately does not support named exports from `.ts` files. Try to have a named export in a file and us that in your `serverless.yml` - it won't work and you won't get any information on why that is the case.

**Comment by pl0xy** [comment]
@arielweinberger yeah, i am jsut about to try using `serverless-esbuild` for my cfworkers project instead of the guide i *was* following since the framework can do cf workers anyway, and reading the docs of esbuild it does most everything webpack can do but it seems  to to it better?

I'll comment again here when i have things set up and (maybe) something working about my experience with it. cause if it can replace `serverless-webpack` for me, that's a good thing, webpack is complex, and kinda bulky, and again, the speed claims of esbuild are, if true, astonishing

**Comment by rgarcianat** [comment]
Hey I've used [serverless-plugin-typescript](https://github.com/serverless/serverless-plugin-typescript), It worked good when I first installed but then I notices that It was a little bit slow, also I notice that sometimes without doing changes to the code it does a double compile.

**Comment by samchungy** [comment]
I may be a little bias here but I picked up `serverless-esbuild` about a year and a half ago but what made me love it was the relatively low config nature of it compared to say `serverless-webpack`. I would say that _most_ lambdas _should_ be relatively small and uncomplicated bits of code which I think it suits. I'm still trying my best to fix and iron out some bits and pieces of the functionality but I think we're at a point where it's fairly stable.

**Comment by pl0xy** [comment]
@samchungy This is what I'm most interested in. I'm an SA, and one the systems I design stick quite stringently to the "single responsibility principle" meaning that there are many lambdas but in one project (an entire health tech start-up's backend) only the "Create User" lambda" was more than 100 lines long, (it had a little extra steps it had to do to verify. each individual package size was always 52kb. (basically, the logger, it's dependencies and that's about it. cause <100lines of code don't take up much space.

So a new builder that can build these many lambda packages even  faster is gonna be really exciting.

I had to build a new PC with 64GB RAM and 16 cores back in 2019 cause my old one literally couldn't build a service with only 20 lambdas lmao.  ofc now that's done on some CI tooll, like code pipelines, but it was early days.

**Comment by mnapoli** [comment]
Thanks all for your responses!

AFAICT serverless-esbuild seems to be the best option because of its speed.

But is that plugin missing (especially useful) features that the webpack or bundle plugins provide? It could be tempting to go with the esbuild plugin, but I'd love to make sure we don't miss out on anything.

**Comment by arielweinberger** [comment]
Webpack does have a small ecosystem of plugins. Module Federation becoming popular for example. In general, it is a very popular choice for bundling and it's very customizable.

However I think for the most common use cases, these might not even be needed/usable in a Lambda setup.

It's a hard decision to make as webpack/snowpack/rollup/vite/esbuild are all very popular and good in their own ways.

So far the most common use case I've used a solid bundler for is using TypeScript and getting faster build times than the official package. Being able to import other file types (e.g. JSON) is supported by pretty much all of them.

I wouldn't mind esbuild :)

**Comment by mnapoli** [comment]
Thanks for these details!

And right, keeping in mind that this is about finding the default plugin (that covers 80% of use cases), it could make sense to go with the simplest and fastest. And users would have the choice to use a more advanced one for advanced needs. So esbuild might be the good option here it seems 🤔

**Comment by arielweinberger** [comment]
Indeed @mnapoli. I think esbuild would be a good _"default"_ fit because it requires no config (or very little for 80% of use cases). Whereas Webpack's configurability is its strength as well as its weakness. I believe that people who need those specific advanced use cases (e.g. module federation, exotic file handlers) that Webpack can offer, will find it easy to `npm install` the serverless webpack plugin and take it from there.

**Comment by coyoteecd** [comment]
To chime in with an opinion, one of the reasons we're sticking to `serverless-webpack` is the fact that esbuild does not support TypeScript decorators. We use `typeorm` to connect to Aurora via Data API, and it relies on decorators for the db entity definitions. A more in-depth discussion, together with a potential solution for `serverless-esbuild` in the form of an esbuild plugin, can be found [here](https://github.com/evanw/esbuild/issues/257).

**Comment by mnapoli** [comment]
For the record I've created a basic TypeScript example in https://github.com/serverless/examples/pull/685 (by "basic" I mean an example that looks exactly like the default AWS Node (JS) example) -> this example uses the esbuild plugin.

There are still plenty of TypeScript examples [in the examples repository](https://github.com/serverless/examples), I'll try to update them one by one (to use esbuild). If one wants to contribute, feel free to send PRs too.

**Comment by mnapoli** [comment]
[How We Sped Up Our TypeScript Serverless Builds Ten Times (70 Lambdas Under 1 Minute)](https://medium.com/@arsenyyankovski/how-we-sped-up-our-typescript-serverless-builds-ten-times-70-lambdas-under-1-minute-f79a925dfe4c) -> they replaced the webpack plugin with the esbuild plugin 😄

**Comment by deathemperor** [comment]
I used serverless-webpack before then switched to serverless-bundle. I'm migrating to serverless-esbuild today thanks to this conversation. Fast build time for both offline and deployments is neat. This also completes our front and back ends with esbuild (our react apps use vite which uses esbuild)

**Comment by astuyve** [comment]
I'd echo a strong vote for `serverless-esbuild`. I've found it's not only faster, but given that hot module replacement isn't strictly necessary in the Serverless environment, the  broader feature set of webpack seems less useful given the context.

tl:dr; `esbuild` satisfies this requirement well:
> And right, keeping in mind that this is about finding the default plugin (that covers 80% of use cases), it could make sense to go with the simplest and fastest.

**Comment by kilakewe** [comment]
I've just been testing the performance difference between serverless-bundle (which we've been using for the past 2 years) and serverless-esbuild for a TS based project with ~50 endpoints across 6 stacks (serverless-compose based). 

We've cut our **build time** to 1/10th of what it was previously, bringing our **deploy** time down from 12 minutes to 5 minutes.
